compiler/rustc_interface/src/util.rs

```diff
@@ -504,15 +513,14 @@ pub fn build_output_filenames(attrs: &[ast::Attribute], sess: &Session) -> Outpu
                 sess.dcx().emit_warn(errors::IgnoringOutDir);
             }
 
-            let out_filestem =
-                out_file.filestem().unwrap_or_default().to_str().unwrap().to_string();
+            let out_filestem = out_file.filestem().unwrap_or_default().to_str().unwrap();
             OutputFilenames::new(
                 out_file.parent().unwrap_or_else(|| Path::new("")).to_path_buf(),
-                crate_name.unwrap_or_else(|| out_filestem.replace('-', "_")),
+                crate_name,
```

Comment:
Indeed, this is a change in behavior. On master, the "crate name" falls back to the file stem of the explicit outfile (`-o`) which is not correct (it doesn't agree with `tcx.crate_name(LOCAL_CRATE)`) (this happens since #114750). Note that this "crate name" is only used when emitting metadata files via `--emit=metadata` namely inside `libCRATE_NAME.rmeta`.

However, I couldn't actually come up with an actual reproducer as `--emit=metadata` doesn't actually emit anything if an `-o OUTPATH` is present **unless** an outpath was explicitly specified via `--emit=metadata=RMETA_OUTPATH`. So it might not be reachable at all and thus have no user-facing impact.

I don't know if `--emit=metadata -o OUTPATH` not emitting anything is intentional. Needs investigation (by me).